neo/SmartContract/Native/Tokens/Nep5Token.cs

```diff
@@ -0,0 +1,139 @@
+            }
+        }
+
+        internal void MintTokens(ApplicationEngine engine, UInt160 account, BigInteger amount)
+        {
+            if (amount.Sign < 0) throw new ArgumentOutOfRangeException(nameof(amount));
+            if (amount.IsZero) return;
+            StorageItem storage = engine.Service.Snapshot.Storages.GetAndChange(CreateAccountKey(account), () => new StorageItem
+            {
+                Value = new TState().ToByteArray()
+            });
+            TState state = new TState();
+            state.FromByteArray(storage.Value);
+            state.Balance += amount;
+            storage.Value = state.ToByteArray();
+            engine.Service.SendNotification(engine, ScriptHash, new StackItem[] { "Transfer", StackItem.Null, account.ToArray(), amount });
+        }
+
+        protected abstract BigInteger TotalSupply(ApplicationEngine engine);
+
+        protected virtual BigInteger BalanceOf(ApplicationEngine engine, UInt160 account)
+        {
+            StorageItem storage = engine.Service.Snapshot.Storages.TryGet(CreateAccountKey(account));
+            if (storage is null) return BigInteger.Zero;
+            Nep5AccountState state = new Nep5AccountState(storage.Value);
+            return state.Balance;
+        }
+
+        protected virtual bool Transfer(ApplicationEngine engine, UInt160 from, UInt160 to, BigInteger amount)
+        {
+            if (engine.Service.Trigger != TriggerType.Application) throw new InvalidOperationException();
+            if (amount.Sign < 0) throw new ArgumentOutOfRangeException(nameof(amount));
+            if (!from.Equals(new UInt160(engine.CurrentContext.CallingScriptHash)) && !engine.Service.CheckWitness(engine, from))
+                return false;
+            ContractState contract_to = engine.Service.Snapshot.Contracts.TryGet(to);
+            if (contract_to?.Payable == false) return false;
+            StorageKey key_from = CreateAccountKey(from);
+            StorageItem storage_from = engine.Service.Snapshot.Storages.TryGet(key_from);
+            if (amount.IsZero)
+            {
+                if (storage_from != null)
+                {
+                    TState state_from = new TState();
+                    state_from.FromByteArray(storage_from.Value);
+                    OnBalanceChanging(engine, from, state_from, amount);
+                }
+            }
+            else
+            {
+                if (storage_from is null) return false;
+                TState state_from = new TState();
+                state_from.FromByteArray(storage_from.Value);
+                if (state_from.Balance < amount) return false;
+                if (from.Equals(to))
+                {
+                    OnBalanceChanging(engine, from, state_from, BigInteger.Zero);
+                }
+                else
+                {
+                    OnBalanceChanging(engine, from, state_from, -amount);
+                    if (state_from.Balance == amount)
```

Comment:
I think this requires `state_from` to be passed as `const` to the OnBalanceChanging... it's reference based on C# rather than copy based, right?
Must ensure no one will change the state_from.Balance on the method above.

Reply:
We need to change `state_from` in `OnBalanceChanging()`.